core/drivers/crypto/stm32/ecc.c

```diff
@@ -0,0 +1,506 @@
+
+static TEE_Result algo_to_pka_cid(uint32_t algo,
+				  enum stm32_pka_curve_id *cid)
+{
+	switch (algo) {
+	case TEE_ALG_ECDSA_P192:
+		*cid = PKA_NIST_P192;
+		break;
+	case TEE_ALG_ECDSA_P224:
+		*cid = PKA_NIST_P224;
+		break;
+	case TEE_ALG_ECDSA_P256:
+		*cid = PKA_NIST_P256;
+		break;
+	case TEE_ALG_ECDSA_P384:
+		*cid = PKA_NIST_P384;
+		break;
+	case TEE_ALG_ECDSA_P521:
+		*cid = PKA_NIST_P521;
+		break;
+	default:
+		EMSG("algorithm %#"PRIx32" not enabled", algo);
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result curve_to_pka_cid(uint32_t curve,
+				   enum stm32_pka_curve_id *cid)
+{
+	switch (curve) {
+	case TEE_ECC_CURVE_NIST_P192:
+		*cid = PKA_NIST_P192;
+		break;
+	case TEE_ECC_CURVE_NIST_P224:
+		*cid = PKA_NIST_P224;
+		break;
+	case TEE_ECC_CURVE_NIST_P256:
+		*cid = PKA_NIST_P256;
+		break;
+	case TEE_ECC_CURVE_NIST_P384:
+		*cid = PKA_NIST_P384;
+		break;
+	case TEE_ECC_CURVE_NIST_P521:
+		*cid = PKA_NIST_P521;
+		break;
+	default:
+		EMSG("curve %#"PRIx32" not enabled", curve);
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_gen_keypair(struct ecc_keypair *key, size_t size_bits)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct stm32_pka_bn d = { };
+	struct stm32_pka_point pk = { };
+	enum stm32_pka_curve_id cid = PKA_LAST_CID;
+	size_t bytes = 0;
+
+	if (!key)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	FMSG("Using PKA");
+	res = curve_to_pka_cid(key->curve, &cid);
+	if (res)
+		return res;
+
+	res = stm32_pka_get_max_size(&bytes, NULL, cid);
+	if (res)
+		return res;
+
+	if (size_bits > bytes * 8 ||
+	    crypto_bignum_num_bytes(key->d) > bytes ||
+	    crypto_bignum_num_bytes(key->x) > bytes ||
+	    crypto_bignum_num_bytes(key->y) > bytes)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	d.val = calloc(1, bytes);
+	d.size = bytes;
+	if (!d.val)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	/* Private key is a random vector */
+	res = crypto_rng_read(d.val, d.size);
+	if (res) {
+		free(d.val);
+		return res;
+	}
+
+	pk.x.val = calloc(1, bytes);
+	pk.x.size = bytes;
+	if (!pk.x.val) {
+		free(d.val);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	pk.y.val = calloc(1, bytes);
+	pk.y.size = bytes;
+	if (!pk.y.val) {
+		free(pk.x.val);
+		free(d.val);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	res = stm32_pka_edac_gen_pubkey(&d, &pk, cid);
+	if (res)
+		goto out;
+
+	res = crypto_bignum_bin2bn(d.val, d.size, key->d);
+	if (res)
+		goto out;
+
+	res = crypto_bignum_bin2bn(pk.x.val, pk.x.size, key->x);
+	if (res)
+		goto out;
+
+	res = crypto_bignum_bin2bn(pk.y.val, pk.y.size, key->y);
+
+out:
+	free(pk.y.val);
+	free(pk.x.val);
+	free(d.val);
+
+	return res;
+}
+
+static TEE_Result sign(uint32_t algo, struct ecc_keypair *key,
+		       const uint8_t *msg, size_t msg_size,
+		       uint8_t *sig, size_t *sig_len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	enum stm32_pka_curve_id cid_from_algo = PKA_LAST_CID;
+	enum stm32_pka_curve_id cid = -1;
```

Comment:
I don't want to use `PKA_LAST_CID` as I'm comparing `cid_from_algo != cid`
I don't want to initialise `cid_from_algo==cid` as it will not catch an error in case of attack jumping the call to `curve_to_pka_cid()` and `algo_to_pka_cid()`

Reply:
After discussion in person, I agreed to remove the protection as OP-TEE seems to not expect this level of protection against attacks